rfcs/0012-external-function.md

```diff
@@ -0,0 +1,163 @@
+---
+feature: external-function-framework
+authors:
+  - "TennyZhuang"
+start_date: "2022/11/04"
+---
+
+# External Function Framework
+
+## Summary
+
+We will introduce a gRPC-based external function framework in the kernel part. **The framework is private to users and may change later**.
+
+## Motivation
+
+The RFC is a preparation RFC for external UDFs with heavy runtime or complicated environments.
+
+The RFC will not answer the following questions:
+
+* How to define the UDFs?
+* How to execute the UDFs?
+
+## Design
+
+### External function protocol
+
+We will add a simple protocol for external function servers to implement:
+
+```protobuf
+message BatchExternalUdfRequest {
+    repeated bytes columns = 1;
+}
+
+message BatchExternalUdfResponse {
```

Comment:
How is the user expected to report error?

Reply:
Not sure, but we can't handle the error elegantly, so just leave it later.